adapters/richaudience/richaudience.go

```diff
@@ -0,0 +1,355 @@
+	richaudienceRequest := richaudienceRequest{}
+
+	raiHeaders := http.Header{}
+
+	raiErrors := make([]error, 0, len(request.Imp))
+
+	var regsExt *openrtb_ext.ExtRegs
+	if request.Regs != nil && request.Regs.Ext != nil {
+		if err := json.Unmarshal(request.Regs.Ext, &regsExt); err != nil {
+			raiErrors = append(raiErrors, &errortypes.BadInput{
+				Message: err.Error(),
+			})
+			return nil, raiErrors
+		}
+	}
+
+	setHeaders(&raiHeaders)
+
+	richaudienceRequest.ID = request.ID
+
+	resImps, err := setImp(request, &richaudienceRequest)
+
+	if err != nil {
+		raiErrors = append(raiErrors, &errortypes.BadInput{
+			Message: err.Error(),
+		})
+		return nil, raiErrors
+	}
+
+	richaudienceRequest.Imp = resImps
+
+	setSite(request, &richaudienceRequest)
+
+	err = setDevice(request, &richaudienceRequest)
+	if err != nil {
+		raiErrors = append(raiErrors, &errortypes.BadInput{
+			Message: err.Error(),
+		})
+		return nil, raiErrors
+	}
+
+	err = setUser(request, &richaudienceRequest, requestInfo)
+	if err != nil {
+		raiErrors = append(raiErrors, &errortypes.BadInput{
+			Message: err.Error(),
+		})
+		return nil, raiErrors
+	}
+
+	//User: consent
+	req, err := json.Marshal(richaudienceRequest)
+	if err != nil {
+		raiErrors = append(raiErrors, &errortypes.BadInput{
+			Message: err.Error(),
+		})
+		return nil, raiErrors
+	}
+
+	requestData := &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    req,
+		Headers: raiHeaders,
+	}
+
+	return []*adapters.RequestData{requestData}, raiErrors
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if responseData.StatusCode == http.StatusBadRequest {
+		err := &errortypes.BadInput{
+			Message: "Unexpected status code: 400. Bad request from publisher. Run with request.debug = 1 for more info.",
+		}
+		return nil, []error{err}
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var bidResp openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &bidResp); err != nil {
+
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: err.Error(),
+		}}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+	for _, seatBid := range response.SeatBid {
+		for _, bid := range seatBid.Bid {
+			b := &adapters.TypedBid{
+				Bid:     &bid,
```

Comment:
There is a subtle bug here. You cannot pass a reference to the bid loop variable outside of the loop. Please see the examples in our new bidder guide: https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html#adapter-code

Reply:
in next commit i have solved